bicytok/distanceMetricFuncs.py

```diff
@@ -1,641 +1,221 @@
-        for optimal_transport, receptor_name, signal_receptor in sorted_results[:10]
-    ]
-    # bar graph
-    receptor_names = [info[0] for info in top_receptor_info]
-    distances = [info[1] for info in top_receptor_info]
-    if ax is not None:
-        ax.bar(range(len(receptor_names)), distances)
-        ax.set_xlabel("Receptor")
-        ax.set_ylabel("Distance")
-        ax.set_title("Top 5 Receptor Distances (2D)")
-        ax.set_xticks(range(len(receptor_names)))
-        ax.set_xticklabels(receptor_names, rotation="vertical")
-
-    return sorted_results
-
-
-def EMD_3D(dataset1, target_cells, ax=None):
+
+    KL_div_vals = np.full(recAbundances.shape[1], np.nan)
+    EMD_vals = np.full(recAbundances.shape[1], np.nan)
+
+    for rec in range(recAbundances.shape[1]):
+        targAbun, offTargAbun = normalize(recAbundances[:, rec], targ, offTarg)
+
+        if np.mean(targAbun) > np.mean(offTargAbun):
+            KL_div_vals[rec] = calculateKLDiv(targAbun, offTargAbun, dims=1)
+            EMD_vals[rec] = stats.wasserstein_distance(
+                targAbun, offTargAbun
+            )  # Consider switching/comparing to ot.emd2_1d
+
+    return KL_div_vals, EMD_vals
+
+
+def KL_EMD_2D(recAbundances, targ, offTarg):
     """
-    returns list of descending EMD values for specified target cell (3 receptors)
-    returns list of descending EMD values for specified target cell (3 receptors)
-    :param dataset1: DataFrame containing the dataset
-    :param target_cells: Target cell type for analysis
-    :param ax: Matplotlib Axes object for plotting (optional)
+    Calculates 2D EMD and KL Divergence between the target and off-target populations of two receptors, across multiple receptors
+    :param recAbundances: an array of receptor abundances across cells (rows) and receptors (columns)
+    :param targ: a binary/boolian vector indicating indices of target cells
+    :param offTarg: a binary/boolian vector indicating indices of off-target cells
     :return:
-    List of tuples (format: (recep1, recep2, recep 3, OT value) containing optimal transport distances and receptor information for 3D analysis
+        KL_div_vals: an array of KL Divergences between the on- and off-target abundances of multiple receptors
+            for each entry the value is calculated for two receptors, so every entry is for a different receptor pair
+            the array is triangular because it is symmetric across the diagonal
+            ???the diagonal values are the 1D distances??? Check this
+        EMD_vals: similar to KL_div_vals but with EMDs
+    """
+
+    # Do conversion factors in functions?
+    # weightDF = convFactCalc(recAbundances) ...
+
+    KL_div_vals = np.empty([recAbundances.shape[1], recAbundances.shape[1]])
+    KL_div_vals[:] = np.nan
+    EMD_vals = np.empty([recAbundances.shape[1], recAbundances.shape[1]])
+    EMD_vals[:] = np.nan
+
+    for rec1 in range(recAbundances.shape[1]):
```

Comment:
I think you can make one loop as `for rec1, rec2 in np.tril_indices_from(recAbundances):`

Reply:
Done